libraries/lib-module-manager/PluginHost.cpp

```diff
@@ -0,0 +1,220 @@
+      }
+   }
+}
+
+PluginHost::PluginHost()
+{
+   FileNames::InitializePathList();
+
+   wxFileName configFileName{ FileNames::Configuration() };
+   auto pConfig = std::make_unique<FileConfig>(
+      AppName, wxEmptyString, configFileName.GetFullPath(),
+      wxEmptyString, wxCONFIG_USE_LOCAL_FILE);
+   pConfig->Init();
+   InitPreferences(std::move(pConfig));
+
+   auto& moduleManager = ModuleManager::Get();
+   moduleManager.Initialize();
+   moduleManager.DiscoverProviders();
+
+   mClient = std::make_unique<IPCClient>(*this);
+}
+
+void PluginHost::OnConnect(IPCChannel& channel) noexcept
+{
+   std::lock_guard lck(mSync);
+   mChannel = &channel;
+}
+
+void PluginHost::OnDisconnect() noexcept
+{
+   Stop();
+}
+
+void PluginHost::OnConnectionError() noexcept
+{
+   Stop();
+}
+
+void PluginHost::OnDataAvailable(const void* data, size_t size) noexcept
+{
+   try
+   {
+      mInputMessageReader.ConsumeBytes(data, size);
+      if(mInputMessageReader.CanPop())
+      {
+         {
+            std::lock_guard lck(mSync);
+            assert(!mRequest);
+            mRequest = mInputMessageReader.Pop();
+         }
+         mRequestCondition.notify_one();
+      }
+   }
+   catch(...)
+   {
+      Stop();
+   }
+}
+
+bool PluginHost::Serve()
+{
+   std::unique_lock lck(mSync);
+   mRequestCondition.wait(lck, [this]{ return !mRunning || mRequest.has_value(); });
+
+   if(!mRunning)
+      return false;
+
+   if(mRequest)
+   {
+      std::optional<wxString> request;
+      mRequest.swap(request);
+
+      lck.unlock();
+      
+      wxString providerId;
+      wxString pluginPath;
+      details::PluginValidationResult result;
+      if(details::ParseRequestString(*request, providerId, pluginPath))
+         Discover(result, providerId, pluginPath);
+      else
+         result.SetError("malformed request string");
+
+      XMLStringWriter xmlWriter;
+      result.WriteXML(xmlWriter);
+
+      lck.lock();
+      if(mChannel)
+         details::PutMessage(*mChannel, xmlWriter);
+   }
+
+   return true;
+}
+
+
+void PluginHost::Stop() noexcept
+{
+   try
+   {
+      {
+         std::lock_guard lck(mSync);//may throw
+         mRunning = false;
+         mChannel = nullptr;
+      }
+   }
+   catch(...)
+   {
+      //If something went wrong with mutex locking we'll try to
+      //awake main thread if it's blocked on conditinal variable.
+      //Attempt to relock the mutex there should throw as well(?)...
+      //which, in turn, will result in std::terminate being called
+   }
+   mRequestCondition.notify_one();
+}
+
+bool PluginHost::Start()
+{
+   const auto cmd = wxString::Format("\"%s\" %s", PlatformCompatibility::GetExecutablePath(), PluginHost::HostArgument);
+
+   auto process = std::make_unique<wxProcess>();
+   process->Detach();
+   if(wxExecute(cmd, wxEXEC_ASYNC, process.get()) != 0)
+   {
+      //process will delete itself upon termination
+      process.release();
+      return true;
+   }
+   return false;
```

Comment:
Maybe it won’t be difficult to build a small executable that links only this library and its dependencies.  Avoid wxModule and wasteful loading of another Audacity with all of its static initializations.

Not essential to making this PR.

Reply:
I was thinking of it too, and there is a downside - the size of application will increase, don't know how much though.